llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -10760,16 +10762,38 @@ SDValue RISCVTargetLowering::lowerMaskedLoad(SDValue Op,
   if (!VL)
     VL = getDefaultVLOps(VT, ContainerVT, DL, DAG, Subtarget).second;
 
-  unsigned IntID =
-      IsUnmasked ? Intrinsic::riscv_vle : Intrinsic::riscv_vle_mask;
+  SDValue Index;
+  if (!IsUnmasked && IsExpandingLoad) {
+    MVT IndexVT = ContainerVT;
+    if (ContainerVT.isFloatingPoint())
+      IndexVT = IndexVT.changeVectorElementTypeToInteger();
+
+    if (Subtarget.isRV32() && IndexVT.getVectorElementType().bitsGT(XLenVT))
+      IndexVT = IndexVT.changeVectorElementType(XLenVT);
```

Comment:
Is it possible to have a <512 x i8> vector where Index would overflow? We might need to increase the size of IndexVT in that case

Reply:
Yes, we can construct such tests like:
```
define <1024 x i8> @test_expandload_v1024i8(ptr %base, <1024 x i1> %mask, <1024 x i8> %passthru) "target-features"="+zvl1024b" {
  %res = call <1024 x i8> @llvm.masked.expandload.v1024i8(ptr align 1 %base, <1024 x i1> %mask, <1024 x i8> %passthru)
  ret <1024 x i8> %res
}
```
But I don't know how to make it legal because we can't just change `IndexVT` to be with element type `MVT::i16` as the changed `IndexVT` will be an illegal type. For example, from `nxv64i8` to `nxv64i16`(illegal).
I tried to split it to small ones but failed. Do you have any thought on fixing this or can we just ignore it?

Reply:
> I tried to split it to small ones but failed.

What failed?

Reply:
> > I tried to split it to small ones but failed.
> 
> What failed?

Well, I may not say it exactly. I just tried to reuse legalizer to split, but it may not be possible. Now I am trying to split it manually in lowering phase.

Reply:
I think the vector shuffle lowering runs into the same issue, it just bails if it can't promote the index to i16

https://github.com/llvm/llvm-project/blob/b1234ddbe2652aa7948242a57107ca7ab12fd2f8/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L5290-L5297

Reply:
I have implemented the splitting operation, which is copied and modified from `DAGTypeLegalizer::SplitVecRes_MLOAD`.